### PR TITLE
Add support for date_unit="D" in read_json and to_json

### LINKED ISSUE #16492: No way with to_json to write only date out of datetime
#### Code Sample, a copy-pastable example if possible

```python
In [2]: pd.Series(pd.to_datetime(['2017-03-15'])).to_csv()
Out[2]: '0,2017-03-15\n'

In [3]: pd.Series(pd.to_datetime(['2017-03-15'])).to_json(date_format='iso')
Out[3]: '{"0":"2017-03-15T00:00:00.000Z"}'
```
#### Problem description

From [this SO comment](https://stackoverflow.com/questions/16176996/keep-only-date-part-when-using-pandas-to-datetime/41718815?noredirect=1#comment75167203_41718815)

By default, ``to_csv()`` drops times of day if they are all midnight. By default, ``to_json`` instead writes dates in epoch, but if passed ``date_format='iso'`` will always write times.

Not sure whether it would be optimal to have ``date_format='iso'`` behave as ``to_csv()``, or rather add a new (for instance) ``date_format='iso_date'``, or maybe even free-form like accepted by ``pd.Series.df.strftime()``, e.g. ``date_format="%Y-%m-%d"``.

#### Expected Output
```python
Out[3]: '{"0":"2017-03-15"}'
```

#### Output of ``pd.show_versions()``

<details>
# Paste the output here pd.shoINSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.7.0-1-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.utf8
LOCALE: it_IT.UTF-8

pandas: 0.20.1
pytest: 3.0.6
pip: 9.0.1
setuptools: None
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: 0.9.2
IPython: 5.1.0.dev
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.2
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.6
lxml: 3.7.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: 0.2.1
w_versions() here

</details>

### PR DESCRIPTION
**Pull request by DavideCanton**
- [x] closes #16492
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This PR extends the `date_unit` parameter of the `pd.read_json` and the `to_json` method of Series and DataFrame methods by adding another allowed value: `"D"`.

This value can be used to serialize and deserialize dates into `YYYY-MM-DD` format.

In the deserialization, I chose 365 as min value for the D unit since all the minimum values represent a year.

The differences with the other supported units in deserialization are:
- this unit is not automatically inferred, I've left the default min to `"s"` to avoid breaking retrocompatibility.
- this unit is not case insensitive

I've already updated unit tests and the documentation. I wasn't sure of where this enhancement should be put in the whatsnew, so feel free to suggest me and I will update it.

All suggestions are welcome, please let me know what I can improve.

### CODE REVIEW

File: pandas/io/json/_json.py [hunk 1]
```diff
@@ -1063,7 +1066,9 @@ def __init__(
         self.dtype = dtype
 
         if date_unit is not None:
-            date_unit = date_unit.lower()
+            # avoid lowercasing "D" but ensure retrocompatibility for other units
+            if date_unit != "D":
```
Comment (mroeschke) [inline pandas/io/json/_json.py:1070]: Why not just put a lowercase D in `_STAMP_UNITS`?

**Comment by mroeschke** [comment]
> In the deserialization, I chose 365 as min value for the D unit since all the minimum values represent a year.

Does this work correctly when calling this during a leap year?

File: pandas/io/json/_json.py (hunk 1 above)
Comment (DavideCanton) [reply pandas/io/json/_json.py:1070]: I wanted to keep the exact same format passed to the underlying function

**Comment by DavideCanton** [comment]
> > In the deserialization, I chose 365 as min value for the D unit since all the minimum values represent a year.
> 
> Does this work correctly when calling this during a leap year?

What do you mean? This threshold is used only when detecting if a number is a suitable value for decoding a timestamp with date unit "D".

**Comment by WillAyd** [comment]
Not an expert in this area but I'm also a little unsure of how accurate this would be. The code seems to assume there is always 24 hours in a day which isn't true when you think about things like DST

Do you know how numpy internally manages going from nanoseconds to day precision?

**Comment by DavideCanton** [comment]
> Not an expert in this area but I'm also a little unsure of how accurate this would be. The code seems to assume there is always 24 hours in a day which isn't true when you think about things like DST
> 
> Do you know how numpy internally manages going from nanoseconds to day precision?

I tried some manipulations and it seems to truncate the time part, at least from my experiments:

```py
In [105]: a = np.array(["2023-03-25T00:23:45.248755"], dtype="datetime64[ns]")

In [106]: b = a + np.array([np.timedelta64(i, 'h') for i in range(30)])

In [107]: b.astype('datetime64[D]')
Out[107]:
array(['2023-03-25', '2023-03-25', '2023-03-25', '2023-03-25',
       '2023-03-25', '2023-03-25', '2023-03-25', '2023-03-25',
       '2023-03-25', '2023-03-25', '2023-03-25', '2023-03-25',
       '2023-03-25', '2023-03-25', '2023-03-25', '2023-03-25',
       '2023-03-25', '2023-03-25', '2023-03-25', '2023-03-25',
       '2023-03-25', '2023-03-25', '2023-03-25', '2023-03-25',
       '2023-03-26', '2023-03-26', '2023-03-26', '2023-03-26',
       '2023-03-26', '2023-03-26'], dtype='datetime64[D]')

In [108]: b - a
Out[108]:
array([              0,   3600000000000,   7200000000000,  10800000000000,
        14400000000000,  18000000000000,  21600000000000,  25200000000000,
        28800000000000,  32400000000000,  36000000000000,  39600000000000,
        43200000000000,  46800000000000,  50400000000000,  54000000000000,
        57600000000000,  61200000000000,  64800000000000,  68400000000000,
        72000000000000,  75600000000000,  79200000000000,  82800000000000,
        86400000000000,  90000000000000,  93600000000000,  97200000000000,
       100800000000000, 104400000000000], dtype='timedelta64[ns]')

In [109]: (b-a).astype('timedelta64[D]')
Out[109]:
array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1], dtype='timedelta64[D]')
```

**Comment by jbrockmendel** [comment]
> Do you know how numpy internally manages going from nanoseconds to day precision?

division/multiplication.  numpy does everything with timezone-naive datetimes.

**Comment by DavideCanton** [comment]
Hi, is there any interest in merging this PR? If so, I can resolve the conflicts with the main branch

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Comment by DavideCanton** [comment]
I didn't update this on purpose, since I got no feedback and had no time to keep it updated. As I said before, if there is any interest in this feature, I'll update it

**Comment by mousedownmike** [comment]
I would have used this feature if it were available.  My use case involves regulatory data that applies to a day of the year (no regard for timezone).  

Having an output option that does not include the time of day would significantly reduce the payload size.  As previously discussed, this would still be an ISO 8601 conformant date format.

I'm subscribed to this issue and happy to provide additional information/support if it will help.

**Comment by DavideCanton** [comment]
Hi @mousedownmike! The feature in this PR was working at the time I created the PR, I was waiting for the review by some maintainer, but it was developed on pandas 1.5.x, so this probably needs some work to adapt all the code to the new major.

I'll probably look into it, just I don't know when, since in this period I'm quite busy, if you have some spare time feel free to work on it!

**Comment by DavideCanton** [comment]
@mroeschke I've updated the code, can you please reopen the PR?

**Comment by DavideCanton** [comment]
@mroeschke why almost all the checks were cancelled?

**Comment by mroeschke** [comment]
Could you merge main and push to this branch again?

**Comment by WillAyd** [comment]
How do we see this aligning with the efforts in https://github.com/pandas-dev/pandas/pull/53757#pullrequestreview-1536170940 ? I think we are developing two different ways to deal with various levels of precision. 

I'm still inclined to go with the other approach as it is more native to the type system, but maybe that doesn't include all of the work here

**Comment by DavideCanton** [comment]
> How do we see this aligning with the efforts in [#53757 (review)](https://github.com/pandas-dev/pandas/pull/53757#pullrequestreview-1536170940) ? I think we are developing two different ways to deal with various levels of precision.
> 
> I'm still inclined to go with the other approach as it is more native to the type system, but maybe that doesn't include all of the work here

It makes sense, since this PR was opened before pandas 2, when things were slightly different.

Said so, I read a bit the linked PR and the issue, and I agree it doesn't include all the support for `unit="D"` since it seems not supported in `read_json`. I'm not very familiar with pandas internals though, so I may have missed something.

Maybe it may make sense to either close this PR and have a fresh start when the other is merged, or rebase this onto the one you provided?

**Comment by mroeschke** [comment]
Looks like https://github.com/pandas-dev/pandas/pull/53757 has been merged. I agree it's probably best to have a fresh PR start relative to that PR so closing.

**Pull request closed without merging**